src/MuladdMacro.jl

```diff
@@ -1,138 +1,184 @@
-        else
-            next_factor1 = Expr(:call, :*, next_operands[1:end-1]...)
-        end
-
-        # create a dot call if both involved operators are dot calls
-        if isdotcall(ex)
-            Expr(:., Base.muladd, Expr(:tuple, next_factor1, next_factor2, last_expr))
-        else
-            Expr(:call, Base.muladd, next_factor1, next_factor2, last_expr)
-        end
+"""
+    sub_to_muladd(ex)
+
+Replace subtraction `ex` by `muladd` if possible.
+"""
+function sub_to_muladd(ex)
+    # Retrieve operands
+    x, y = args(ex)
+
+    # Modify subtraction if possible
+    dotcall = isdotcall(ex)
+    if ismul(y, dotcall)
+        y₁, y₂ = splitargs(y)
+        return newmuladd(:(-$y₁), y₂, x, dotcall)
+    elseif ismul(x, dotcall)
+        return newmuladd(splitargs(x)..., :(-$y), dotcall)
     end
+
+    return ex
 end
-to_muladd(ex) = ex
 
 """
-    isoperation(ex, op::Symbol)
+    isdotcall(ex)
 
-Determine whether `ex` is a call of operation `op`.
+Determine whether `ex` is a dot call.
 """
-isoperation(ex::Expr, op::Symbol) =
-    ex.head == :call && !isempty(ex.args) && ex.args[1] == op
-isoperation(ex, op::Symbol) = false
+isdotcall(ex) =
+    (@capture(ex, (f_)(__)) && startswith(string(f), '.')) ||
+    @capture(ex, (_).(__))
 
 """
-    isdotcall(ex[, op])
+    issum(ex)
 
-Determine whether `ex` is a dot call and, in case `op` is specified, whether it calls
-operator `op`.
+Determine whether `ex` is a sum.
 """
-isdotcall(ex::Expr) = !isempty(ex.args) &&
-    (ex.head == :. ||
-     (ex.head == :call && !isempty(ex.args) && first(string(ex.args[1])) == '.'))
-isdotcall(ex) = false
+issum(ex) = @capture(ex, +(__) | .+(__) | (+).(__))
 
-isdotcall(ex::Expr, op::Symbol) = isdotcall(ex) &&
-    (ex.args[1] == op || ex.args[1] == Symbol('.', op))
-isdotcall(ex, op::Symbol) = false
+"""
+    issub(ex)
 
+Determine whether `ex` is a subtraction.
 """
-    isaddition(ex)
+issub(ex) = @capture(ex, -(_, _) | .-(_, _) | (-).(_, _))
 
-Determine whether `ex` is an expression of an addition.
 """
-isaddition(ex) = isoperation(ex, :+) || isdotcall(ex, :+)
+    ismul(ex, dot::Bool)
 
+Determine whether expression `ex` is a multiplication that is dotted if
+`dot` is `true` and not dotted otherwise.
 """
-    operands(ex)
+function ismul(ex, dot::Bool)
+    if dot
+        return @capture(ex, .*(__) | (*).(__))
+    else
+        return @capture(ex, *(__))
+    end
+end
 
-Return arguments of function call in `ex`.
 """
-function operands(ex::Expr)
-    if ex.head == :. && length(ex.args) == 2 && typeof(ex.args[2]) <: Expr
-        ex.args[2].args
+    newmuladd(x, y, z, dot::Bool)
+
+Return expression `(muladd).(x, y, z)` if `dot` is `true` and
+`muladd(x, y, z)` otherwise.
+"""
+function newmuladd(x, y, z, dot::Bool)
+    # Quoting seems to be required for the @. macro to work
+    if dot
+        :(($(Meta.quot(Base.muladd))).($x, $y, $z))
     else
-        ex.args[2:end]
+        :($(Meta.quot(Base.muladd))($x, $y, $z))
     end
```

Comment:
I'm not sure why quoting is needed here and whether that's the correct way to handle the application of `@.` macros. But it was the only way I could get it to work besides expanding `@.` macros immediately which in my opinion should not be required (or at least feels a bit suboptimal).